packages/indexer-common/src/indexer-management/allocations.ts

```diff
@@ -1148,4 +1157,73 @@ export class AllocationManager {
     return isDeploymentWorthAllocatingTowards(logger, subgraphDeployment, indexingRules)
       .toAllocate
   }
+
+  // Calculates the balance (GRT delta) of a single Action.
+  async resolveActionDelta(action: Action): Promise<ResolvedAction> {
+    let unallocates = BigNumber.from(0)
+
+    // Handle allocations
+    const allocates = BigNumber.from(action.amount ?? 0)
+
+    // Handle unallocations
+    if (action.type === ActionType.UNALLOCATE || action.type === ActionType.REALLOCATE) {
+      // ensure this Action have a valid allocationID
+      if (action.allocationID === null || action.allocationID === undefined) {
+        throw Error(
+          `SHOULD BE UNREACHABLE: Unallocate or Reallocate action must have an allocationID field: ${action}`,
+        )
+      }
+      // Fetch the allocation on chain to inspect its amount
+      const allocation = await this.networkMonitor.allocation(action.allocationID)
+      if (!allocation) {
+        throw indexerError(
+          IndexerErrorCode.IE063,
+          `Action validation failed: No active allocation found on chain with id '${action.allocationID}' found`,
+        )
+      }
+      /* We intentionally don't check if the allocation is active now because it will be checked
+       * later when we prepare the transaction */
+      unallocates = unallocates.add(allocation.allocatedTokens)
+    }
+
+    const balance = allocates.sub(unallocates)
+    return {
+      action,
+      allocates,
+      unallocates,
+      balance,
+    }
+  }
+
+  async validateActionBatchFeasibilty(batch: Action[]): Promise<Action[]> {
+    const logger = this.logger.child({ function: 'validateActionBatch' })
+    logger.debug(`Validating action batch`, { size: batch.length })
+
+    // Validate stake feasibility
+    const freeStake = await this.contracts.staking.getIndexerCapacity(this.indexer)
+    const mapper = async (action: Action) => this.resolveActionDelta(action)
+    const resolvedBatch = await pMap(batch, mapper)
+    const batchDelta: BigNumber = resolvedBatch
+      .map((resolvedAction) => resolvedAction.balance)
+      .reduce((a, b) => a.add(b))
+    const newBalance = batchDelta.add(freeStake)
```

Comment:
I think this should be `freeStake.sub(batchDelta)` 🤔 

From what I can tell the `resolveActionDelta()` function returns a positive value if a specific action is going to be allocating more than it is unallocating; meaning more stake will be used moving forward than currently used. As such a batch of actions will have a positive batchDelta if that batch is going to be using GRT from the freeStake pool --> batchDelta should be subtracted from freeStake to get the new freeStake value post batch execution.

Reply:
Fixed in https://github.com/graphprotocol/indexer/pull/579/commits/360b33ef86f211c77fcf158e75f85d139b90a0f5